v2/federation/watch.go

```diff
@@ -20,16 +26,50 @@ type BundleWatcher interface {
 
 	// OnUpdate is called when a bundle has been updated. If a bundle is
 	// fetched but has not changed from the previously fetched bundle, OnUpdate
-	// will not be called.
+	// will not be called. This function is called synchronously from WatchBundle
+	// so all the time this function takes to finish is added to the time taken
+	// by fetching the bundle.
 	OnUpdate(*spiffebundle.Bundle)
 
 	// OnError is called if there is an error fetching the bundle from the
-	// endpoint.
+	// endpoint. This function is called synchronously from WatchBundle
+	// so all the time this function takes to finish is added to the time taken
+	// by fetching the bundle.
 	OnError(err error)
 }
 
 // WatchBundle watches a bundle on a bundle endpoint. It returns when the
 // context is canceled, returning ctx.Err().
 func WatchBundle(ctx context.Context, trustDomain spiffeid.TrustDomain, url string, watcher BundleWatcher, options ...FetchOption) error {
-	panic("not implemented")
+	if watcher == nil {
+		return federationErr.New("watcher cannot be nil")
+	}
+
+	var lastBundle *spiffebundle.Bundle
+	d := watcher.NextRefresh(0)
+	timer := time.NewTimer(d)
+	defer timer.Stop()
+	for {
+		bundle, err := fetchBundleCallback(ctx, trustDomain, url, options)
+		switch {
+		// Context was canceled when fetching bundle, so to avoid
+		// more calls to FetchBundle (because the timer could be expired at
+		// this point) we return now.
+		case ctx.Err() == context.Canceled:
+			return ctx.Err()
+		case err != nil:
+			watcher.OnError(err)
+		case !lastBundle.Equal(bundle):
+			watcher.OnUpdate(bundle)
+			lastBundle = bundle
+		}
+
+		select {
+		case <-timer.C:
+			d = watcher.NextRefresh(0)
```

Comment:
NextRefresh should be passed the refresh hint from the bundle, if available.

Reply:
Oh, I didn't see that the bundle had a function to retrieve the refresh hint :grimacing: